File: implementations/typescript/src/backends/ontology.ts
/**
 * @file ontology.ts
 * @description Ontology backend interface for the UPP protocol.
 *
 * Defines the {@link OntologyBackend} interface for accessing ontology
 * metadata, label definitions, and server information.
 *
 * @module
 */

import type { LabelDefinition } from "../models/index.js";

// ─────────────────────────────────────────────────────────────────────────────
// OntologyBackend Interface
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Interface for pluggable ontology backends.
 *
 * An ontology backend provides access to label definitions and the list
 * of available ontologies.
 */
export interface OntologyBackend {
  /**
   * Return all label definitions for the server's ontology.
   *
   * @returns Label definitions.
   */
  getLabels(): Promise<LabelDefinition[]>;

  /**
   * Return the ontology identifier for this server instance.
   *
   * @returns Ontology identifier string (e.g., `"user/v1"`).
   */
  getOntology(): Promise<string>;
}

File: implementations/python/src/upp/backends/ontology.py
"""UPP Ontology Backend Protocol.

Defines the :class:`OntologyBackend` protocol for accessing ontology
metadata, label definitions, and server information.
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from upp.models.labels import LabelDefinition

__all__ = [
    "OntologyBackend",
]


@runtime_checkable
class OntologyBackend(Protocol):
    """Protocol for pluggable ontology backends.

    An ontology backend provides access to label definitions, the
    ontology identifier, and server information for the ``upp/info``
    operation.
    """

    def get_labels(self) -> list[LabelDefinition]:
        """Return all label definitions for the ontology.

        Returns:
            A list of :class:`LabelDefinition` objects.
        """
        ...

    def get_version(self) -> str:
        """Return the ontology identifier for this server instance.

        Returns:
            The ontology identifier string (e.g., ``"user/v1"``).
        """
        ...
